cabi/src/proguard.rs

```diff
@@ -77,3 +77,129 @@ ffi_fn! {
         Ok(SymbolicProguardMappingView::as_rust(view).has_line_info())
     }
 }
+
+use proguard::{ProguardMapper, ProguardMapping, StackFrame};
+
+/// Represents a Java Stack Frame.
+#[repr(C)]
+pub struct SymbolicJavaStackFrame {
+    pub klass: SymbolicStr,
+    pub method: SymbolicStr,
+    pub file: SymbolicStr,
+    pub line: usize,
+}
+
+/// The result of remapping a Stack Frame.
+#[repr(C)]
+pub struct SymbolicProguardRemapResult {
+    pub frames: *mut SymbolicJavaStackFrame,
+    pub len: usize,
+}
+
+pub struct OwnedProguardMapper<'s> {
+    _source: String,
+    mapping: ProguardMapping<'s>,
+    mapper: ProguardMapper<'s>,
+}
+
+/// Represents a proguard mapper.
+pub struct SymbolicProguardMapper;
+
+impl ForeignObject for SymbolicProguardMapper {
+    type RustObject = OwnedProguardMapper<'static>;
+}
+
+ffi_fn! {
+    /// Creates a proguard mapping view from a path.
+    unsafe fn symbolic_proguardmapper_open(
+        path: *const c_char
+    ) -> Result<*mut SymbolicProguardMapper> {
+        let path = CStr::from_ptr(path).to_str()?;
+        let source = std::fs::read_to_string(path)?;
+        let mapping = ProguardMapping::new(std::mem::transmute(source.as_str()));
+        let mapper = ProguardMapper::new(mapping.clone());
+
+        let proguard = OwnedProguardMapper { _source: source, mapping, mapper };
+
+        Ok(SymbolicProguardMapper::from_rust(proguard))
+    }
+}
+
+ffi_fn! {
+    /// Frees a proguard mapping view.
+    unsafe fn symbolic_proguardmapper_free(mapper: *mut SymbolicProguardMapper) {
+        SymbolicProguardMapper::drop(mapper);
+    }
+}
+
+ffi_fn! {
+    /// Remaps a Stack Frame.
+    unsafe fn symbolic_proguardmapper_remap_frame(
+        mapper: *const SymbolicProguardMapper,
+        class: *const SymbolicStr,
+        method: *const SymbolicStr,
+        line: usize,
+    ) -> Result<SymbolicProguardRemapResult> {
+        let mapper = &SymbolicProguardMapper::as_rust(mapper).mapper;
+        let frame = StackFrame::new((*class).as_str(), (*method).as_str(), line);
+
+        let mut frames: Vec<_> = mapper.remap_frame(&frame).map(|frame| {
+            SymbolicJavaStackFrame {
+                klass: frame.class().into(),
+                method: frame.method().into(),
```

Comment:
This `SymbolicStr::from` only borrows the `&str`, but the underlying `Cow` will be dropped after the `map` closure has finished. You should clone the string, therefore.

```suggestion
                klass: frame.class().to_owned().into(),
                method: frame.method().to_owned().into(),
```

The same applies to `file` below, as well as `symbolic_proguardmapper_remap_class`. I'm a little bit curious why this does not segfault in the tests.

Reply:
> I'm a little bit curious why this does not segfault in the tests.

The borrowed string lives for the complete lifetime of the mapper. But yes, better be safe here.